sqlmesh/core/engine_adapter/base.py

```diff
@@ -1145,18 +1204,102 @@ def scd_type_2(
             columns_to_types = self.columns(target_table)
         if not columns_to_types:
             raise SQLMeshError(f"Could not get columns_to_types. Does {target_table} exist?")
-        if updated_at_name not in columns_to_types:
+        if not unique_key:
+            raise SQLMeshError("unique_key must be provided for SCD Type 2")
+        if check_columns and updated_at_name:
+            raise SQLMeshError(
+                "Cannot use both `check_columns` and `updated_at_name` for SCD Type 2"
+            )
+        if check_columns and updated_at_as_valid_from:
+            raise SQLMeshError(
+                "Cannot use both `check_columns` and `updated_at_as_valid_from` for SCD Type 2"
+            )
+        if execution_time_as_valid_from and not check_columns:
+            raise SQLMeshError(
+                "Cannot use `execution_time_as_valid_from` without `check_columns` for SCD Type 2"
+            )
+        if updated_at_name and updated_at_name not in columns_to_types:
             raise SQLMeshError(
                 f"Column {updated_at_name} not found in {target_table}. Table must contain an `updated_at` timestamp for SCD Type 2"
             )
+
         unmanaged_columns = [
             col for col in columns_to_types if col not in {valid_from_name, valid_to_name}
         ]
-        valid_from_start = (
-            updated_at_name
-            if updated_at_as_valid_from
-            else self._to_utc_timestamp("1970-01-01 00:00:00+00:00")
-        )
+        # If a star is provided, we include all unmanaged columns in the check.
+        # This unnecessarily includes unique key columns but since they are used in the join, and therefore we know
+        # they are equal or not, the extra check is not a problem and we gain simplified logic here.
+        # If we want to change this, then we just need to check the expressions in unique_key and pull out the
+        # column names and then remove them from the unmanaged_columns
+        if check_columns and check_columns[0] == exp.Star():
+            check_columns = [exp.to_column(col) for col in unmanaged_columns]
+        execution_ts = self._to_utc_timestamp(to_ts(execution_time))
+        update_valid_from_start: t.Union[str, exp.Expression]
+        if updated_at_as_valid_from:
+            if not updated_at_name:
+                raise SQLMeshError(
+                    "Cannot use `updated_at_as_valid_from` without `updated_at_name` for SCD Type 2"
+                )
+            update_valid_from_start = updated_at_name
+        elif execution_time_as_valid_from:
+            update_valid_from_start = execution_ts
+        else:
+            update_valid_from_start = self._to_utc_timestamp("1970-01-01 00:00:00+00:00")
+        insert_valid_from_start = execution_ts if check_columns else updated_at_name
+        if check_columns:
+            updated_row_filter = " OR ".join(
```

Comment:
could be safer to use sqlglot builder instead of raw string manipulation, in case col has bad identifiers in them

```python
col.neq(tcol).or_(tcol.is(exp.null()).and_(col.is(exp.null()).not_())
for col in check_columns

Reply:
Addressed here: https://github.com/TobikoData/sqlmesh/pull/1997/commits/b61e09420ae8676fca71433b5a6fe60db1845adb